mono/metadata/icall.c

```diff
@@ -8733,111 +8734,113 @@ mono_lookup_icall_symbol (MonoMethod *m)
+#define ICALL_SIG_TYPE_bool     (&mono_defaults.boolean_class)
+#define ICALL_SIG_TYPE_boolean  (&mono_defaults.boolean_class)
+#define ICALL_SIG_TYPE_double   (&mono_defaults.double_class)
+#define ICALL_SIG_TYPE_float    (&mono_defaults.single_class)
+#define ICALL_SIG_TYPE_int      (&mono_defaults.int_class)
+#define ICALL_SIG_TYPE_int16    (&mono_defaults.int16_class)
+#define ICALL_SIG_TYPE_int32    (&mono_defaults.int32_class)
+#define ICALL_SIG_TYPE_int8     (&mono_defaults.sbyte_class)
+#define ICALL_SIG_TYPE_long     (&mono_defaults.int64_class)
+#define ICALL_SIG_TYPE_obj      (&mono_defaults.object_class)
+#define ICALL_SIG_TYPE_object   (&mono_defaults.object_class)
+#define ICALL_SIG_TYPE_ptr      (&mono_defaults.int_class) // ?
+#define ICALL_SIG_TYPE_string   (&mono_defaults.string_class)
+#define ICALL_SIG_TYPE_uint16   (&mono_defaults.uint16_class)
+#define ICALL_SIG_TYPE_uint32   (&mono_defaults.uint32_class)
+#define ICALL_SIG_TYPE_uint8    (&mono_defaults.byte_class)
+#define ICALL_SIG_TYPE_ulong    (&mono_defaults.uint64_class)
+#define ICALL_SIG_TYPE_void     (&mono_defaults.void_class)
+
+#define ICALL_SIG_TYPES_1(a) 		  	ICALL_SIG_TYPE_ ## a,
+#define ICALL_SIG_TYPES_2(a, b) 	  	ICALL_SIG_TYPES_1 (a            ) ICALL_SIG_TYPES_1 (b)
+#define ICALL_SIG_TYPES_3(a, b, c) 	  	ICALL_SIG_TYPES_2 (a, b         ) ICALL_SIG_TYPES_1 (c)
+#define ICALL_SIG_TYPES_4(a, b, c, d) 	  	ICALL_SIG_TYPES_3 (a, b, c      ) ICALL_SIG_TYPES_1 (d)
+#define ICALL_SIG_TYPES_5(a, b, c, d, e)	ICALL_SIG_TYPES_4 (a, b, c, d   ) ICALL_SIG_TYPES_1 (e)
+#define ICALL_SIG_TYPES_6(a, b, c, d, e, f)	ICALL_SIG_TYPES_5 (a, b, c, d, e) ICALL_SIG_TYPES_1 (f)
+#define ICALL_SIG_TYPES_7(a, b, c, d, e, f, g)	ICALL_SIG_TYPES_6 (a, b, c, d, e, f) ICALL_SIG_TYPES_1 (g)
+#define ICALL_SIG_TYPES_8(a, b, c, d, e, f, g, h) ICALL_SIG_TYPES_7 (a, b, c, d, e, f, g) ICALL_SIG_TYPES_1 (h)
+
+#define ICALL_SIG_TYPES(n, types) ICALL_SIG_TYPES_ ## n types
+
+// A scheme to make these const would be nice.
+// For example, if types[] could remain indirect, or
+// an offset within mono_defaults.
+static struct {
+#define ICALL_SIG(n, xtypes) 			\
+	struct {				\
+		MonoMethodSignature sig;	\
+		MonoClass** types [n];		\
+	} ICALL_SIG_NAME (n, xtypes);
+ICALL_SIGS
+	MonoMethodSignature end;
+} mono_icall_signatures = {
+#undef ICALL_SIG
+#define ICALL_SIG(n, types) { { \
+	0,			/* ret */ \
+	n,			/* param_count */ \
+	-1,			/* sentinelpos */ \
+	0,			/* generic_param_count */ \
+	MONO_ICALL_SIGNATURE_CALL_CONVENTION, \
+	0,			/* hasthis */ \
+	0, 			/* explicit_this */ \
+	1, 			/* pinvoke */ \
+	0, 			/* is_inflated */ \
+	0,			/* has_type_parameters */ \
+}, { ICALL_SIG_TYPES (n, types) } }, /* params and ret */
+ICALL_SIGS
+};
 
-	g_strfreev (parts);
+#undef ICALL_SIG
+#define ICALL_SIG(n, types) MonoMethodSignature * const ICALL_SIG_NAME (n, types) = &mono_icall_signatures.ICALL_SIG_NAME (n, types).sig;
+ICALL_SIGS
+#undef ICALL_SIG
 
-	mono_image_lock (corlib);
```

Comment:
I wonder if the signatures should be further hashed/single-instanced based on their eventual non-string representation.
To some extent that can be done statically, `obj` vs. `object`.
But they historically were not in either case -- just based on input string -- so this is progress.